Implement sign in via facebook

allenwq opened the pull request.

lowjoel commented on config/initializers/devise.rb:
```diff
@@ -236,6 +236,7 @@
   # Add a new OmniAuth provider. Check the wiki for more information on setting
   # up on your models and hooks.
   # config.omniauth :github, 'APP_ID', 'APP_SECRET', scope: 'user,public_repo'
+  config.omniauth :facebook, ENV['FACEBOOK_APP_ID'], ENV['FACEBOOK_APP_SECRET']
```
prefix with OMNIAUTH

lowjoel commented on app/models/user.rb:
```diff
@@ -1,6 +1,7 @@
 # Represents a user in the application. Users are shared across all instances.
 class User < ActiveRecord::Base
   include UserAuthenticationConcern
+  include UserOmniauthConcern
```
Include this from the authentication concern.

allenwq replied:
The method overriding(`#new_with_session`) is not working after I include this in the authentication concern.

I'm lean towards to just merge two concerns.

lowjoel replied:
I don't see a spec that checks for this.

lowjoel replied:
You can fix this by including the UserOmniauthConcern at the end of the UserAuthenticationConcern included block. This is because including in the module includes the module before devise. so devise is called and not you.

anyway there should be a model spec that checks for this.

allenwq replied:
Thanks! I'll try this.

lowjoel commented on app/controllers/user/omniauth_callbacks_controller.rb:
```diff
@@ -0,0 +1,27 @@
+      sign_in_and_redirect @user, event: :authentication
+      set_flash_message(:notice, :success, kind: t('.facebook')) if is_navigational_format?
+    else
+      session['devise.facebook_data'] = auth
```
How would this affect the registration form?

lowjoel commented on app/controllers/user/omniauth_callbacks_controller.rb:
```diff
@@ -0,0 +1,27 @@
+      set_flash_message(:notice, :success, kind: t('.facebook')) if is_navigational_format?
+    else
+      session['devise.facebook_data'] = auth
+      redirect_to new_user_registration_url
```
_path

allenwq commented on app/models/concerns/user_omniauth_concern.rb:
```diff
@@ -0,0 +1,52 @@
+    # This method is a override of devise {#new_with_session}
+    def new_with_session(params, session)
+      super.tap do |user|
+        facebook_data = session['devise.facebook_data']
```
@lowjoel The data is used here.

And the data starting with devise was because:
>In case the user is not persisted, we store the OmniAuth data in the session. Notice we store this data using "devise." as key namespace. This is useful because Devise removes all the data starting with "devise." from the session whenever a user signs in, so we get automatic session clean up.

REF: https://github.com/plataformatec/devise/wiki/OmniAuth:-Overview#facebook-example

lowjoel replied:
how does this scale for adding multiple authentication methods later on?

allenwq replied:
I can imagine that in future the code would be like:
```
if facebook_data && facebook_data['info']
  user.update_from_facebook_data(facebook_data)
elsif twitter_data
  user.update_from_twitter_data(twitter_data)
...
```

lowjoel replied:
start writing the method dispatch now.

lowjoel replied:
actually should this be handled by the user class, or the user instance?

allenwq replied:
Actually a private method in this concern will do, I moved to another function just because rubocop complains.

allenwq replied:
What's the correct way of declaring a private method in concern ? (don't need to include by user)

lowjoel replied:
no such thing as private method in a module, when it's mixed into the class, the class would have the private method.

allenwq replied:
So I should declare a class method instead  ?

allenwq replied:
My "private" means the method is only accessible in the module itself.

lowjoel commented on app/models/concerns/user_authentication_concern.rb:
```diff
@@ -2,6 +2,7 @@ module UserAuthenticationConcern
   extend ActiveSupport::Concern
 
   included do
+    include UserOmniauthConcern
```
include within the concern, not inside the included.

allenwq replied:
OK, this is also the part that I was considering

lowjoel commented on app/models/concerns/user_omniauth_concern.rb:
```diff
@@ -0,0 +1,52 @@
+  module ClassMethods
+    # Find or create a user from omniauth data.
+    #
+    # @param [Hash] auth The data with omniauth info.
```
if this is a hash, why is line 14 using dot syntax?

allenwq replied:
my bad, should be `OmniAuth::AuthHash`

lowjoel commented on app/models/concerns/user_omniauth_concern.rb:
```diff
@@ -0,0 +1,52 @@
+    end
+
+    # A convenience method that receives both parameters and session to initialize a user.
+    # This method is a override of devise {#new_with_session}
```
find the full identifier and link to it.

lowjoel commented on app/models/concerns/user_omniauth_concern.rb:
```diff
@@ -0,0 +1,52 @@
+module UserOmniauthConcern
```
I think this concern is not properly structured for multiple providers in future.

allenwq replied:
Only `#new_with_session` need to consider different providers, check my comments there ?

lowjoel replied:
Ideally I'll just include more concerns like this and it should automatically work with more providers. This allows me to keep the implementation details of each provider separate.

There's two levels of abstraction in this concern. One deals with the user. One deals with the social network. They should live separately.

lowjoel commented on app/controllers/user/omniauth_callbacks_controller.rb:
```diff
@@ -0,0 +1,28 @@
+    auth = request.env['omniauth.auth']
+    @user = User.from_omniauth(auth)
+    if @user.persisted?
+      sign_in_and_redirect @user, event: :authentication
```
`sign_in_and_redirect()`

lowjoel commented on app/models/concerns/user_omniauth_concern.rb:
```diff
@@ -0,0 +1,52 @@
+    # @param [Hash] data The data with omniauth info.
+    # @return [User] The user with its attributes from the given info.
+    def update_user_with_auth_data(user, data)
+      user.email = data.info.email
```
user.assign_attributes

lowjoel commented on spec/controllers/user/omniauth_callbacks_controller_spec.rb:
```diff
@@ -0,0 +1,45 @@
+        expect(identity).to be_present
+      end
+
+      it 'creates a user with the info from facebook'do
```
information

lowjoel commented on app/models/concerns/user_omniauth_concern.rb:
```diff
@@ -0,0 +1,34 @@
+    # Find or create a user from omniauth data.
+    #
+    # @param [OmniAuth::AuthHash] auth The data with omniauth info.
+    # @return [User] The user found or created from omniauth data.
```
This function returns nil also.

allenwq replied:
This function won't return nil, either User or non persisted User.

lowjoel replied:
wait, right.

lowjoel commented on app/models/concerns/user_omniauth_concern.rb:
```diff
@@ -0,0 +1,34 @@
+      return identity.user if identity
+
+      User.create do |user|
+        user.assign_attributes(name: auth.info.name, email: auth.info.email,
```
what if the returned authentication info doesn't have the email address? users have the option of not releasing the email address.

allenwq replied:
Then the user won't be created, he will be redirected to the registration url

lowjoel replied:
That’s not obvious from the flow.

allenwq replied:
That why I have `if user.persisted`, if not persisted means the information is not enough. Then he will be redirected and required to fill in the missing information himself.

lowjoel replied:
I think if this is called `find_or_create_by_identity` then it works like the normal find_or_create_by_X method, which would return an unpersisted object if save could not be done.

lowjoel commented on app/models/concerns/user_omniauth_concern.rb:
```diff
@@ -0,0 +1,34 @@
+    # This is a override of `Devise::Models::Registerable::ClassMethods#new_with_session`
+    def new_with_session(params, session)
+      super.tap do |user|
+        user.instance_users.build
```
we are building this at every sign in? Even if not using omniauth?

lowjoel replied:
wait, no, this is used when registering a user. this is independent of whether the user is signing in using omniauth or not? Why does `from_omniauth` then need to explicitly build an instance user?

allenwq replied:
This callback will only be called when user was redirected to the registration form. It's different from ``from_omniauth` `

lowjoel replied:
That’s not what my interpretation of the Devise source code seems to be…

allenwq replied:
I haven't checked their source code for this part before. 

But just now I saw here seems to be the only part that this method was called:
https://github.com/plataformatec/devise/blob/1a0192201b317d3f1bac88f5c5b4926d527b1b39/app/controllers/devise/registrations_controller.rb#L100

allenwq replied:
`from_omniauth` directly created the user and skipped the devise registrable controller actions, so I need to especially build an instance_user for it.

lowjoel commented on app/controllers/user/omniauth_callbacks_controller.rb:
```diff
@@ -0,0 +1,28 @@
+    if @user.persisted?
+      sign_in_and_redirect(@user, event: :authentication)
+      set_flash_message(:notice, :success,
+                        kind: t('user.omniauth_callbacks.facebook.kind')) if is_navigational_format?
```
use the lazy format.

allenwq replied:
What is lazy format ?

lowjoel replied:
t(‘.facebook.kind’)

allenwq replied:
Initially I was using it, but turns out that I18n-tasks have complains: `en.user.omniauth_callbakcs.sign_in_with_facebook.facebook.kind` missing, note that it treats `sign_in_with_facebook` as an action.

I think it's a issue of them.

allenwq replied:
I guess the same reason here:
https://github.com/Coursemology/coursemology2/blob/master/app/controllers/course/user_registrations_controller.rb#L50

lowjoel replied:
oh, right, I didn't see the action.

lowjoel commented on spec/controllers/user/omniauth_callbacks_controller_spec.rb:
```diff
@@ -0,0 +1,45 @@
+          facebook_data.info.email = nil
+        end
+
+        it 'does not create any user'do
```
missing space before do

lowjoel commented on app/models/concerns/user_omniauth_concern.rb:
```diff
@@ -0,0 +1,34 @@
+    # @param [OmniAuth::AuthHash] auth The data with omniauth info.
+    # @return [User] The user found or created from omniauth data.
+    def from_omniauth(auth)
+      identity = User::Identity.find_by(provider: auth.provider, uid: auth.uid)
```
isn't this whole method just a `find_or_initialize_by`?

allenwq replied:
Why need initialize ?

lowjoel replied:
because if you cannot find the identity you create a new user?

allenwq replied:
Oh you mean the whole `from_omniauth` method's naming ?

lowjoel replied:
yes, the whole function looks like a find_or_initialize_by (or find_or_create_by)

lowjoel replied:
...seems like find_or* cannot handle associations.

maybe call this `find_or_create_by_identity()`?

allenwq replied:
Yes it's a better name, but we should not override active_record method `find_or_create_by`?  any other alternative names ?

allenwq replied:
How about `find_or_create_by_omniauth` ? because its parameter is omniauth info, not identity

lowjoel replied:
that works, guess it's less ambiguous

allenwq replied:
:+1: Thanks, definitely a better name than the example code.

lowjoel commented on app/models/concerns/user_omniauth_concern.rb:
```diff
@@ -0,0 +1,34 @@
+  include UserOmniauthFacebookConcern
+
+  included do
+    devise :omniauthable
```
don't you have to specify the providers?

allenwq replied:
No need, For this I debugged into devise's source code, they can automatically know the providers from the configuration in devise initialisers.

lowjoel replied:
okay, then that's good.

lowjoel commented:
I get the impression this is a mash of example code and self-written code... can this be reviewed again for coherence?

allenwq commented:
@lowjoel I've updated and removed things relate to `instance_users` in this PR, I'll do that in a separate PR for both Sign in and Sign up.

Please see whether it's clear to you now, and of course I'll add more tests later.

lowjoel commented on app/models/concerns/user_omniauth_facebook_concern.rb:
```diff
@@ -0,0 +1,17 @@
+      super.tap do |user|
+        facebook_data = session['devise.facebook_data']
+        if facebook_data && (info = facebook_data['info'])
+          user.assign_attributes(name: info['name'], email: info['email'])
```
should this be done after calling super, or super should be called with modified params? This needs to be tested, please.

lowjoel replied:
(unit tested)

lowjoel replied:
wait, it is tested. okay, sure.

I think this should not matter if it is assigned after super, since it is a `new` and not a `create`?

allenwq replied:
Yep, it's just `new`

lowjoel commented on app/models/concerns/user_omniauth_concern.rb:
```diff
@@ -0,0 +1,31 @@
+    # Find the user from omniauth data.
+    #
+    # @param [OmniAuth::AuthHash] auth The data with omniauth info.
+    # @return [User|nil] The user founded or nil.
```
found

allenwq commented:
@lowjoel Updated

allenwq commented on app/models/concerns/user_omniauth_concern.rb:
```diff
@@ -0,0 +1,25 @@
+    #
+    # @param [OmniAuth::AuthHash] auth The data with omniauth info.
+    # @return [User] The user found or created from omniauth data.
+    def from_omniauth(auth)
```
Hound got huge delays recently...

On Wednesday, July 29, 2015, Hound <notifications@github.com> wrote:

> In app/models/concerns/user_omniauth_concern.rb
> <https://github.com/Coursemology/coursemology2/pull/441#discussion_r35766658>
> :
>
> > @@ -0,0 +1,25 @@
> > +module UserOmniauthConcern
> > +  extend ActiveSupport::Concern
> > +  include UserOmniauthFacebookConcern
> > +
> > +  included do
> > +    devise :omniauthable
> > +  end
> > +
> > +  module ClassMethods
> > +    # Find or create a user from omniauth data.
> > +    #
> > +    # @param [OmniAuth::AuthHash] auth The data with omniauth info.
> > +    # @return [User] The user found or created from omniauth data.
> > +    def from_omniauth(auth)
>
> Assignment Branch Condition size for from_omniauth is too high. [16.06/15]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Coursemology/coursemology2/pull/441/files#r35766658>.
>
